### PR TITLE
Enhanced from_transformers to support class name extraction

### PR DESCRIPTION
**Pull request by jeslinpjames**
# Description

Issue #1043

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

Here is the  [Colab link](https://colab.research.google.com/drive/1xFjqajv-CU57SMF3r8niGaO2IjiGqJou?usp=sharing) with the changes made to the `from_transformers()` function.

## Any specific deployment considerations

For example, documentation changes, usability, usage/costs, secrets, etc.

## Docs

-   [x] Docs updated? What were the changes:
- The documentation for the `from_transformers` function has been updated to include the new parameter `id2label`.

### CODE REVIEW

File: supervision/detection/core.py [hunk 1]
```diff
@@ -388,19 +388,31 @@ def from_mmdetection(cls, mmdet_results) -> Detections:
         )
 
     @classmethod
-    def from_transformers(cls, transformers_results: dict) -> Detections:
+    def from_transformers(
+        cls, transformers_results: dict, id2label: Dict[int, str]
```
Comment (SkalskiP) [inline supervision/detection/core.py:392]: Could we make `id2label` `Optional` and make it `None` by default?

File: supervision/detection/core.py [hunk 2]
```diff
@@ -388,19 +388,31 @@ def from_mmdetection(cls, mmdet_results) -> Detections:
...
+        class_ids = transformers_results["labels"].cpu().numpy().astype(int)
+        class_names = np.array([id2label[class_id] for class_id in class_ids])
         return cls(
-            xyxy=transformers_results["boxes"].cpu().numpy(),
-            confidence=transformers_results["scores"].cpu().numpy(),
-            class_id=transformers_results["labels"].cpu().numpy().astype(int),
+            xyxy=boxes,
+            confidence=scores,
+            class_id=class_ids,
+            data={CLASS_NAME_DATA_FIELD: class_names},
```
Comment (SkalskiP) [inline supervision/detection/core.py:415]: And only add `CLASS_NAME_DATA_FIELD` field to `data` if `id2label` is not `None`.

**Review by SkalskiP — CHANGES_REQUESTED** [review]

**Comment by SkalskiP** [comment]
Hi @jeslinpjames 👋🏻 I only left two minor comments.

**Comment by jeslinpjames** [comment]
Hey @SkalskiP 👋🏻 Thanks for the review! I'll handle those minor adjustments immediately.

**Comment by SkalskiP** [comment]
Hi @jeslinpjames 👋🏻 Thanks a lot for the update. We've also been simultaneously working on another feature related to this part of the code. Could you resolve the conflicts that have arisen?

File: supervision/detection/core.py [hunk 3]
```diff
@@ -404,18 +404,50 @@ def from_transformers(
...
+        masks = transformers_results["masks"].cpu().detach().numpy().astype(bool)
         return cls(
-            xyxy=boxes,
-            confidence=scores,
-            class_id=class_ids,
-            data=data,
+            xyxy=mask_to_xyxy(masks),
+            mask=masks,
+            confidence=transformers_results["scores"].cpu().detach().numpy(),
+            class_id=transformers_results["labels"].cpu().detach().numpy().astype(int),
```
Comment (SkalskiP) [inline supervision/detection/core.py:450]: It looks like you forgot to add a data field in this line.

**Review by SkalskiP — CHANGES_REQUESTED** [review]

**Comment by SkalskiP** [comment]
@jeslinpjames I left one more comment. :)

**Comment by jeslinpjames** [comment]
Hey @SkalskiP, I've gone ahead and resolved the conflicts that arose in the `form_transformers` function. Would you mind reviewing it to ensure that the changes are okay?

**Comment by jeslinpjames** [comment]
Hey @SkalskiP I've committed the changes. Could you please take a look and let me know if any further adjustments are needed?

File: supervision/detection/core.py [hunk 4]
```diff
@@ -389,11 +389,18 @@ def from_mmdetection(cls, mmdet_results) -> Detections:
...
+        cls, transformers_results: dict, id2label: Optional[Dict[int, str]] = None
+    ) -> Detections:
         """
         Creates a Detections instance from object detection or segmentation
         [transformer](https://github.com/huggingface/transformers) inference result.
 
+        Args:
+            transformers_results (dict): A dictionary containing the scores, labels, and boxes for an image
+                as predicted by the Transformers model.
+            id2label (Dict[int, str]): A dictionary mapping class IDs to class names.
```
Comment (LinasKo) [inline supervision/detection/core.py:402]: The docstring should also flag `id2label` as `Optional`.

Similar function: `from_azure_analyze_image`

File: supervision/detection/core.py [hunk 5]
```diff
@@ -421,22 +428,25 @@ def from_transformers(cls, transformers_results: dict) -> Detections:
...
                 confidence=transformers_results["scores"].cpu().detach().numpy(),
-                class_id=transformers_results["labels"]
-                .cpu()
-                .detach()
-                .numpy()
-                .astype(int),
+                class_id=class_ids,
+                data=data,
             )
         masks = transformers_results["masks"].cpu().detach().numpy().astype(bool)
```
Comment (LinasKo) [inline supervision/detection/core.py:443]: While we're at it, let's make the code more defensive.

I've reviewed the [image_processing_detr.py](https://github.com/huggingface/transformers/blob/fed27ffc7ec62837dca9bbfc83442eb3678ee026/src/transformers/models/detr/image_processing_detr.py#L744) in `transformers` and it turns out, with instance or panoptic segmentation `transformers_results` might not have `boxes`, `masks` or `labels`.

Let's add an `elif "masks" in transformers_results:` here and at the end - `raise NotImplementedError("Only object detection and semantic segmentation results are supported")`.

**Review by LinasKo — CHANGES_REQUESTED** [review]
Hi @jeslinpjames 👋🏻

Well done with the implementation - I have very

**Comment by SkalskiP** [comment]
Hi @jeslinpjames 👋🏻 @LinasKo added some good comments. I would be very grateful if you could incorporate these changes, and then we'll be ready to merge.

**Comment by jeslinpjames** [comment]
@LinasKo Thanks for the feedback! I've made the requested changes to the docstring and added the defensive code for masks in `transformers_results`. Could you please review it to ensure everything is okay or if any further adjustments are needed? Thanks!

**Comment by SkalskiP** [comment]
@LinasKo, let me know if we are ready ;)

**Review by LinasKo — APPROVED** [review]
Looks good to me!
Happy to merge in.

**Comment by SkalskiP** [comment]
Hi @jeslinpjames and @LinasKo 👋🏻 I made small docs improvements and tested the code. Looks good!

**Review by SkalskiP — APPROVED** [review]
